### PR TITLE
Safer config saving

### LINKED ISSUE #3165: nvda settings randomly resetting
**Reported by camlorn on 2013-04-17 13:24**
This has happened twice in 3 days.  Nvda will randomly forget _all_ of my settings, including that nvda has been run on my computer before, re displaying the start up dialogs and resetting everything.  Obviously, this is...problematic.
What do I need to do, besides (if I can) comment with how to reliably reproduce it?
The only changes in my setup since this has started happening are the enabling of the us-dvorak keyboard layout.
I'm on Version: master-9098,fb41602, and continually update as soon as new development snapshots come out.  I first saw this on Monday the 15th.
The easiest fix would probably be making a backup of all settings at startup, of some sort.  I'm not really sure why this is happening, but both times were while the system was being very sluggish--maybe some sort of interrupted file writing?
**Blocked by #667**

### PR DESCRIPTION
**Pull request by feerrenrut**
Fixes #3165 

This supersedes #5916. Rather than use [osreplace ](https://pypi.python.org/pypi/pyosreplace/0.1) use `MoveFileExW` called with `ctypes`

Generally the approach is to lessen the chance that ALL config is lost if NVDA is interrupted during the saving process. This is done by first writing to a temporary file and then using this to replace the target config file.

There is a chance that some recently changed configuration will not be saved, if the saving process is interrupted while the temp file is being written. However this will no longer result in ALL config being lost.

### CODE REVIEW

File: source/fileUtils.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#fileUtils.py
+#A part of NonVisual Desktop Access (NVDA)
+#Copyright (C) 2006-2016 NV Access Limited
```
Comment (jcsteh) [inline source/fileUtils.py]: This should instead be:

`+#Copyright (C) 2017 NV Access Limited, Bram Duvigneau`

File: source/fileUtils.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#This file is covered by the GNU General Public License.
+#See the file COPYING for more details.
+import os
+import ctypes
+from contextlib import contextmanager
+from tempfile import NamedTemporaryFile
+from logHandler import log
+
+@contextmanager
+def FaultTolerantFile(name):
```
Comment (jcsteh) [inline source/fileUtils.py]: Please add a docstring with details on what this is/does and maybe example usage. The Python term for one of these `with` things is "context manager".

File: source/fileUtils.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	dirpath, filename = os.path.split(name)
+	with NamedTemporaryFile(dir=dirpath, prefix=filename, suffix='.tmp', delete=False) as f:
+		log.debug(f.name)
+		yield f
+		f.flush()
+		os.fsync(f)
+		f.close()
+		MOVEFILE_REPLACE_EXISTING = 1
```
Comment (jcsteh) [inline source/fileUtils.py]: Please move this constant outside the function.

**Review by jcsteh — APPROVED** [review]

**Comment by Brian1Gaff** [comment]
I could have used this yesterday, when a crash during shut down of windows 
resulted in nvda seeming to start like a new install!
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal email to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
----- Original Message -----

**Pull request merged**